apisix/plugins/google-logging.lua

```diff
@@ -0,0 +1,298 @@
+        retry_delay = { type = "integer", minimum = 0, default = 1 },
+        buffer_duration = { type = "integer", minimum = 1, default = 60 },
+        inactive_timeout = { type = "integer", minimum = 1, default = 10 },
+        batch_max_size = { type = "integer", minimum = 1, default = 100 },
+    },
+    oneOf = {
+        { required = { "auth_config" } },
+        { required = { "auth_file" } },
+    },
+}
+
+
+-- remove stale objects from the memory after timer expires
+local function remove_stale_objects(premature)
+    if premature then
+        return
+    end
+
+    for key, batch in ipairs(buffers) do
+        if #batch.entry_buffer.entries == 0 and #batch.batch_to_process == 0 then
+            core.log.warn("removing batch processor stale object, route id:", tostring(key))
+            buffers[key] = nil
+        end
+    end
+
+    stale_timer_running = false
+end
+
+
+local function send_to_google(oauth, entries)
+    local http_new = http.new()
+    local res, err = http_new:request_uri(oauth.entries_uri, {
+        ssl_verify = false,
+        method = "POST",
+        body = core.json.encode({
+            entries = entries,
+            partialSuccess = false,
+        }),
+        headers = {
+            ["Content-Type"] = "application/json",
+            ["Authorization"] = "Bearer " .. oauth:get_access_token(),
+        },
+    })
+
+    if err then
+        return nil, "failed to write log to google, ", err
+    end
+
+    if res.status ~= 200 then
+        return nil, res.body
+    end
+
+    return res.body
+end
+
+
+local function get_auth_config(config)
+    local err
+    local auth_config = {}
+    if config.auth_config then
+        auth_config = config.auth_config
+    end
+
+    if config.auth_file then
+        local file_content
+        file_content, err = core.utils.get_file(config.auth_file)
+        if file_content then
+            auth_config = core.json.decode(file_content)
+        end
+    end
+
+    if err then
+        return nil, err
+    end
+
+    return auth_config
+end
+
+
+local function get_logger_buffer(conf, ctx)
+    local oauth_client = google_oauth:new(auth_config_cache)
+
+    local process = function(entries)
+        return send_to_google(oauth_client, entries)
+    end
+
+    local config = {
+        name = conf.name or plugin_name,
+        retry_delay = conf.retry_delay,
+        batch_max_size = conf.batch_max_size,
+        max_retry_count = conf.max_retry_count,
+        buffer_duration = conf.buffer_duration,
+        inactive_timeout = conf.inactive_timeout,
+        route_id = ctx.var.route_id,
+        server_addr = ctx.var.server_addr,
+    }
+
+    local buffer, err = batch_processor:new(process, config)
+
+    if not buffer then
+        return nil, "error when creating the batch processor: " .. err
+    end
+
+    return buffer
+end
+
+
+local function get_utc_timestamp()
+    ngx_update_time()
+    local now = tostring(ngx_now())
```

Comment:
We can use math.ceil to get the second & millisecond?

Reply:
yes, updated